### PR TITLE
[RLlib] Fix DQN RLModule forward methods such that they can handle dict spaces

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Description

We don't natively build encoders for dict spaces and so we don't account for them in the forward method of the DQN rlm.
This is an issue because users may still want to use encoder configs for dictionaries or they may want to override DQNRLModule.build_encoder etc.

This PR makes a fix and introduces testing for different types of forward passes, observations spaces and configurations for the DQN RL Module.

### CODE REVIEW

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+torch, nn = try_import_torch()
+
+
+# Custom encoder, config and catalog to test Dict observation spaces.
+# RLlib does not build encoders for Dict observation spaces out of the box so we define our own.
+class DictFlattenEncoder(nn.Module):
+    def __init__(self, obs_space, output_dim=64):
+        super().__init__()
+        total_dim = sum(
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/dqn/tests/test_dqn_rl_module.py:26]: For nested composite spaces this won't work, Gymnasium has [flatdim](https://gymnasium.farama.org/api/spaces/utils/#gymnasium.spaces.utils.flatdim) for this but I'm worried that we don't use Gymnasium's flatten function so the resulting spaces might be mismatched. 
Therefore, this is more a note for the future

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        )
+        self.net = nn.Sequential(
+            nn.Linear(total_dim, output_dim),
+            nn.ReLU(),
+        )
+
+    def forward(self, inputs):
+        obs = inputs[Columns.OBS]
+        flat_obs = torch.cat(
+            [obs[k].reshape(obs[k].shape[0], -1) for k in sorted(obs.keys())],
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/dqn/tests/test_dqn_rl_module.py:37]: Does this work for discrete inputs?

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+
+# Observation space definitions.
+OBS_SPACES = {
+    "box": Box(low=-1.0, high=1.0, shape=(8,), dtype=np.float32),
+    "image": Box(low=0, high=255, shape=(64, 64, 3), dtype=np.uint8),
+    "dict": Dict(
+        {
+            "sensors": Box(low=-1.0, high=1.0, shape=(4,), dtype=np.float32),
+            "position": Box(low=-10.0, high=10.0, shape=(3,), dtype=np.float32),
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/dqn/tests/test_dqn_rl_module.py:67]: Add discrete space for observation

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        if (
+            forward_method == "train"
+        ):  # forward train needs batching, exploration and inference don't
+            module.train()
+            # Create a batch first
+            batch_size = 4
+            obs_list = [obs_space.sample() for _ in range(batch_size)]
+            next_obs_list = [obs_space.sample() for _ in range(batch_size)]
+
+            obs_batch = tree.map_structure(lambda *x: np.stack(x, axis=0), *obs_list)
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/dqn/tests/test_dqn_rl_module.py]: We can add `np.stack(x, axis=0, dtype=np.float32)` to remove the following two lines

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Overall looks good, minor changes and a question about working with Discrete observations

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dqn/tests/test_dqn_rl_module.py:37]: Why not? Should be 1 hot encoded, mhh?

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dqn/tests/test_dqn_rl_module.py:26]: Yea. `DictFlattenEncoder` is just for testing here. Flattening all elements of the dict can be counterproductive if these are, for example, images. I think for any more involved obs space we should not just autogenerate some model under the hood.

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dqn/tests/test_dqn_rl_module.py:67]: Done!

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py (hunk 1 above)
Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dqn/tests/test_dqn_rl_module.py:26]: I agree that for models that are too complex, we shouldn't allow autogeneration

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dqn/tests/test_dqn_rl_module.py:37]: Yes, thats a good point that rationally for discrete inputs they should be one-hotted and we wouldn't get this problem but what happens if a discrete is used? Can we provide a helpful error?

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM

File: rllib/algorithms/dqn/tests/test_dqn_rl_module.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dqn/tests/test_dqn_rl_module.py:26]: Makes sense. Let's not do it for this test file though.
